backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/pporder/process/WEBUI_PP_Order_PrintLabel.java

```diff
@@ -107,38 +111,56 @@ protected String doIt()
 				.streamByIds(selectedRowIds)
 				.collect(ImmutableList.toImmutableList());
 
-		final Set<HuId> huIds = new HashSet<HuId>();
+		final Set<HuId> huIds = new HashSet<>();
 
 		for (final PPOrderLineRow row : selectedRows)
 		{
 			final PPOrderLineType type = row.getType();
-				
-				if (type.isMainProduct() && row.isReceipt())
+
+			if (type.isMainProduct() && row.isReceipt())
+			{
+				final ImmutableList<PPOrderLineRow> includedRows = row.getIncludedRows();
+				includedRows.stream()
+						.filter(ppOrderLineRow -> ppOrderLineRow.getType().isHUOrHUStorage())
+						.map(PPOrderLineRow::getHuId)
+						.filter(Objects::nonNull)
+						.forEach(huIds::add);
+			}
+			else if (type.isHUOrHUStorage())
+			{
+				final HuId huId = row.getHuId();
+
+				if (huId != null && isMainProductHu(huId))
 				{
-					final ImmutableList<PPOrderLineRow> includedRows = row.getIncludedRows();
-					includedRows.stream()
-							.filter(ppOrderLineRow -> ppOrderLineRow.getType().isHUOrHUStorage())
-							.map(PPOrderLineRow::getHuId)
-							.filter(Objects::nonNull)
-							.forEach(huIds::add);
+					huIds.add(huId);
 				}
+			}
 		}
 
 		return huIds;
 	}
-	
+
+	private boolean isMainProductHu(final HuId huId)
```

Comment:
basically, this method matches all HUs/included HUs for received for a finished good (main product).
right?

Reply:
Yup.